### PR TITLE
refactor(internal/jimm/jimm.go): introduces a jimm constructor

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

Introduces a constructor for jimm in internal/jimm. With the lightest touch possible atm.

drive-by fix: the db.Database type holds a `migrated` (uint32) field. however we were passing the database around by value and every time we copied it, whoever tried to use it would get a `CodeUpgradeInProgress` error. In this PR we pass a pointer to the database around, hopefully avoiding further migration voes.

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: cmd/jimmsrv/service/service.go [hunk 1]
```diff
@@ -388,32 +405,33 @@ func NewService(ctx context.Context, p Params) (*Service, error) {
...
-		s.jimm.Dialer = jimm.CacheDialer(s.jimm.Dialer)
+		s.jimmParameters.Dialer = jimm.CacheDialer(s.jimmParameters.Dialer)
 	}
 
 	if _, err := url.Parse(p.DashboardFinalRedirectURL); err != nil {
 		return nil, errors.E(op, err, "failed to parse final redirect url for the dashboard")
 	}
 
-	rebacBackend, err := rebac_admin.SetupBackend(ctx, &s.jimm)
+	// instanciate jimm
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]:
```suggestion
	// instantiate jimm
```

File: cmd/jimmsrv/service/service.go [hunk 2]
```diff
@@ -194,14 +194,15 @@ type Params struct {
 
 // A Service is the implementation of a JIMM server.
 type Service struct {
-	jimm jimm.JIMM
+	jimmParameters jimm.Parameters
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]: Why do the params need to go on the struct? I assume you're doing it to pass around to some functions that were looking at s.jimm but we could turn those methods into functions instead and pass in just the bits they need?

File: cmd/jimmsrv/service/service.go [hunk 3]
```diff
@@ -388,32 +405,33 @@ func NewService(ctx context.Context, p Params) (*Service, error) {
...
+		s.jimmParameters.Dialer = jimm.CacheDialer(s.jimmParameters.Dialer)
 	}
 
 	if _, err := url.Parse(p.DashboardFinalRedirectURL); err != nil {
 		return nil, errors.E(op, err, "failed to parse final redirect url for the dashboard")
 	}
 
-	rebacBackend, err := rebac_admin.SetupBackend(ctx, &s.jimm)
+	// instanciate jimm
+	j, err := jimm.New(s.jimmParameters)
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]:
```suggestion
	s.jimm, err = jimm.New(s.jimmParameters)
```

File: cmd/jimmsrv/service/service.go [hunk 4]
```diff
@@ -509,11 +527,11 @@ func (s *Service) setupDischarger(p Params) (*discharger.MacaroonDischarger, err
 func (s *Service) setupSessionStore(ctx context.Context, sessionSecret []byte) (*pgstore.PGStore, error) {
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]: Should we change this to a function that accepts jimmParams? Then at least we don't need jimmParams to stick around on the service struct. Or, at the risk of overcomplicating things, we could have a separate struct called `serviceBuilder`. Just some ideas to avoid keeping jimmparams around longer than needed.

File: cmd/jimmsrv/service/service.go [hunk 5]
```diff
@@ -258,6 +275,10 @@ func (s *Service) MonitorResources(ctx context.Context) {
 
 // OpenFGACleanup starts a goroutine that cleans up any orphaned tuples from OpenFGA.
 func (s *Service) OpenFGACleanup(ctx context.Context, trigger <-chan time.Time) error {
+	if s.jimm == nil {
+		zapctx.Error(ctx, "jimm not created")
+		return errors.E("jimm not created")
+	}
```
Comment (SimoneDutto) [inline cmd/jimmsrv/service/service.go]: Aren't these s.jimm == nil a good place where we can `panic`. 
I feel like not having configured jimm before starting these services is not a recoverable state

File: internal/jimm/jimm.go [hunk 6]
```diff
@@ -41,16 +42,12 @@ var (
...
-// connections therefore the JIMM object itself does not contain any per-
-// request state.
-type JIMM struct {
+type Parameters struct {
 	// Database is the database used by JIMM, this provides direct access
 	// to the data store. Any client accessing the database directly is
 	// responsible for ensuring that the authenticated user has access to
 	// the data.
-	Database db.Database
+	Database *db.Database
```
Comment (SimoneDutto) [inline internal/jimm/jimm.go]: instead of changing this which required a lot of manual work to update things all around.
Couldn't you change this https://github.com/canonical/jimm/blob/436ba45558dd987bb28d93bb7d1e49bb6dea1eeb/internal/db/db.go#L30 to a pointer?

File: internal/jimm/jimm.go [hunk 7]
```diff
@@ -41,16 +42,12 @@ var (
 	}
 )
 
-// A JIMM provides the business logic for managing resources in the JAAS
-// system. A single JIMM instance is shared by all concurrent API
-// connections therefore the JIMM object itself does not contain any per-
-// request state.
-type JIMM struct {
+type Parameters struct {
```
Comment (SimoneDutto) [inline internal/jimm/jimm.go]: godoc please

File: internal/jimm/jimm.go [hunk 8]
```diff
@@ -193,6 +144,92 @@ type OAuthAuthenticator interface {
...
+
+	return j, nil
+}
+
+// A JIMM provides the business logic for managing resources in the JAAS
+// system. A single JIMM instance is shared by all concurrent API
+// connections therefore the JIMM object itself does not contain any per-
+// request state.
+type JIMM struct {
+	Parameters
```
Comment (SimoneDutto) [inline internal/jimm/jimm.go:265]: since we are embedding the struct, it's not `Parameters` but more `Services`. Because we don't start the service in the constructor, we rather take the service reference and store it in the JIMM struct.

File: internal/jimm/jimm.go [hunk 9]
```diff
@@ -193,6 +144,92 @@ type OAuthAuthenticator interface {
...
+	if err != nil {
+		return nil, err
+	}
+	j.RoleManager = roleManager
+	for _, option := range options {
+		option(j)
+	}
+
+	return j, nil
+}
```
Comment (SimoneDutto) [inline internal/jimm/jimm.go]: i thought this method would have been a method where you receive some parameters (ex. db connection string), and you create the `Database, Openfgaclient`.
So we'll end up with a central point where we validate jimm's config, start the jimm services and return the well defined jimm struct.

File: internal/jimm/admin_test.go [hunk 10]
```diff
@@ -38,10 +45,19 @@ func TestLoginDevice(t *testing.T) {
 func TestGetDeviceSessionToken(t *testing.T) {
 	c := qt.New(t)
 	pollingChan := make(chan string, 1)
+	ofgaClient, _, _, err := jimmtest.SetupTestOFGAClient(c.Name())
```
Comment (kian99) [inline internal/jimm/admin_test.go]: Why did this test get an OpenFGA client?

File: internal/jimm/cloud.go [hunk 11]
```diff
@@ -160,7 +160,7 @@ func (j *JIMM) AddCloudToController(ctx context.Context, user *openfga.User, con
 		return errors.E(op, err)
 	}
 
-	if err := validateCloudRegion(ctx, &j.Database, user, cloud, controllerName); err != nil {
+	if err := validateCloudRegion(ctx, j.Database, user, cloud, controllerName); err != nil {
```
Comment (kian99) [inline internal/jimm/cloud.go:163]: It would've been good to keep the mechanical changes to a separate commit/PR if possible.

File: internal/jimm/jimm.go (hunk 7 above)
Comment (kian99) [inline internal/jimm/jimm.go]: Missing godoc

**Review by kian99 — COMMENTED** [review]
Looks good with some comments

File: internal/jimm/jimm.go (hunk 9 above)
Comment (kian99) [reply internal/jimm/jimm.go]: I think it's correct like this, the JIMM method sets up the business logic, so when we have more "managers", the JIMM struct won't need any of the fields in Params, it will just take the pieces of params like the DB, OpenFGA client, pubSub, etc and pass them to the relevant constructors for the managers. Whoever calls jimm.New() decides what they want to use as a DB, as an auth service, etc (even if you can't mock the DB or auth service today).

File: internal/jimm/jimm.go (hunk 6 above)
Comment (alesstimec) [reply internal/jimm/jimm.go]: but then we'd need a constructor for db.Database.

File: cmd/jimmsrv/service/service.go (hunk 5 above)
Comment (alesstimec) [reply cmd/jimmsrv/service/service.go]: isn't returning an error better than panicking?

File: internal/jimm/jimm.go (hunk 8 above)
Comment (alesstimec) [reply internal/jimm/jimm.go:265]: constructors have parameters. and the constructor will then instantiate all the managers/service provided by JIMM.

File: cmd/jimmsrv/service/service.go (hunk 2 above)
Comment (alesstimec) [reply cmd/jimmsrv/service/service.go]: because we have many methods that just create "stuff" (e.g. oauth authenticator) needed by jimm. and all that "stuff" gets put into the parameters and then passed on to the constructor

File: internal/jimm/admin_test.go (hunk 10 above)
Comment (alesstimec) [reply internal/jimm/admin_test.go]: jimm constructor needs an openfga client, otherwise it cannot start the role manager

File: internal/jimm/cloud.go (hunk 11 above)
Comment (alesstimec) [reply internal/jimm/cloud.go:163]: absolutely.. my bad

File: cmd/jimmsrv/service/service.go (hunk 4 above)
Comment (alesstimec) [reply cmd/jimmsrv/service/service.go]: that is a very good point.. changing

File: cmd/jimmsrv/service/service.go [hunk 12]
```diff
@@ -213,6 +217,10 @@ func (s *Service) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 // monitoring all changes to models. WatchControllers finishes when the
 // given context is canceled, or there is a fatal error watching models.
 func (s *Service) WatchControllers(ctx context.Context) error {
+	if s.jimm == nil {
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]: I wouldn't add all these checks. If jimm is nil and this panics we should see that from our tests. If it's happening in production I'd rather have the stack trace of where we tried to use a nil pointer, because this is an unrecoverable error. With this change all we get is an error messsage, "jimm not created" which is arguably worse because we can't see where it happened.

File: cmd/jimmsrv/service/service.go [hunk 13]
```diff
@@ -499,21 +562,17 @@ func (s *Service) setupDischarger(p Params) (*discharger.MacaroonDischarger, err
...
-func (s *Service) setupSessionStore(ctx context.Context, sessionSecret []byte) (*pgstore.PGStore, error) {
+func (s *Service) setupSessionStore(ctx context.Context, sessionSecret []byte, db *db.Database) (*pgstore.PGStore, error) {
 	const op = errors.Op("setupSessionStore")
 
-	if s.jimm.CredentialStore == nil {
-		return nil, errors.E(op, "credential store is not configured")
-	}
-
-	sqlDb, err := s.jimm.Database.DB.DB()
+	sqlDb, err := db.DB.DB()
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go:559]: Can we improve this at the moment?

File: cmd/jimmsrv/service/service_test.go [hunk 14]
```diff
@@ -82,24 +86,32 @@ func TestDefaultService(t *testing.T) {
...
 	c.Assert(err, qt.ErrorMatches, "jimm cannot start without a credential store")
 }
 
 func TestAuthenticator(t *testing.T) {
 	c := qt.New(t)
 
+	ctx, s := gosvc.NewService(context.Background(), os.Interrupt, syscall.SIGTERM)
+
+	var cancel func()
+	ctx, cancel = context.WithCancel(ctx)
```
Comment (kian99) [inline cmd/jimmsrv/service/service_test.go]: Why are we creating a cancel function here but not in all the other tests?

File: internal/jimm/jimm.go [hunk 15]
```diff
@@ -41,16 +42,13 @@ var (
 	}
 )
 
-// A JIMM provides the business logic for managing resources in the JAAS
-// system. A single JIMM instance is shared by all concurrent API
-// connections therefore the JIMM object itself does not contain any per-
-// request state.
-type JIMM struct {
+// Parameters holds parameters for the JIMM type.
```
Comment (kian99) [inline internal/jimm/jimm.go]: Let's better explain this. 
```suggestion
// Parameters holds the services and static fields passed to the jimm.New() constructor.
// You can provide mock implementations of certain services where necessary for dependency injection.
```

File: internal/jimm/jimm.go [hunk 16]
```diff
@@ -193,6 +145,92 @@ type OAuthAuthenticator interface {
...
+	// RenameRole renames a role in JIMM's DB.
+	RenameRole(ctx context.Context, user *openfga.User, uuid, newName string) error
+	// ListRoles returns a list of roles known to JIMM.
+	// `match` will filter the list fuzzy matching role's name or uuid.
+	ListRoles(ctx context.Context, user *openfga.User, pagination pagination.LimitOffsetPagination, match string) ([]dbmodel.RoleEntry, error)
+	// CountRoles returns the number of roles that exist.
+	CountRoles(ctx context.Context, user *openfga.User) (int, error)
+}
+
+// Option modifies JIMM returned by the constructor.
```
Comment (kian99) [inline internal/jimm/jimm.go]: Again I think it would be good to explain why this is useful.
```suggestion
// Option modifies the JIMM returned by the constructor.
// This is useful to create mock domain objects, e.g. replace the real RoleManager with a mock 
// object in tests.
```

File: internal/jimm/jimm.go [hunk 17]
```diff
@@ -193,6 +145,92 @@ type OAuthAuthenticator interface {
...
+type Option func(*JIMM)
+
+// RoleManagerOption replaces the role manager in the returned JIMM with the specified one.
+var RoleManagerOption = func(rm RoleManager) Option {
+	return func(j *JIMM) {
+		j.RoleManager = rm
+	}
+}
+
+// New returns a new instance of JIMM based on the provided parameters and options.
```
Comment (kian99) [inline internal/jimm/jimm.go]:
```suggestion
// New returns a new instance of JIMM.
// See [Option] and [Parameters] to better understand how to perform dependency injection.
// Primitives like the dialer or authentication service can be mocked at a low level, alternatively
// top business layer objects like the RoleManager can be mocked instead.
```

File: cmd/jimmsrv/service/service.go [hunk 18]
```diff
@@ -289,66 +313,50 @@ func (s *Service) AddCleanup(f func() error) {
 }
 
 // NewService creates a new Service using the given params.
-func NewService(ctx context.Context, p Params) (*Service, error) {
+func NewService(ctx context.Context, p Params, svc *service.Service) (*Service, error) {
```
Comment (kian99) [inline cmd/jimmsrv/service/service.go]: Can we refactor this so that we don't need to pass in a service.Service here.

File: cmd/jimmsrv/service/service.go (hunk 12 above)
Comment (SimoneDutto) [reply cmd/jimmsrv/service/service.go]: i agree, i'd rather panic when creating the service and avoid these checks.

File: cmd/jimmsrv/service/service.go [hunk 19]
```diff
@@ -289,66 +313,50 @@ func (s *Service) AddCleanup(f func() error) {
...
 	const op = errors.Op("NewService")
 
 	s := new(Service)
-	s.mux = chi.NewRouter()
-
-	s.mux.Use(chimiddleware.RequestLogger(&logger.HTTPLogFormatter{}))
-	s.mux.Use(middleware.MeasureHTTPResponseTime)
 
+	jimmParameters := jimm.Parameters{
```
Comment (SimoneDutto) [inline cmd/jimmsrv/service/service.go:302]: i would still call the services. But I won't die on this hill

File: internal/jimm/jimm.go [hunk 20]
```diff
@@ -193,6 +145,92 @@ type OAuthAuthenticator interface {
...
+func New(p Parameters, options ...Option) (*JIMM, error) {
```
Comment (SimoneDutto) [inline internal/jimm/jimm.go]: does it make sense to validate here that our parameters(services) are not nil? So we can remove the checks in the getters

**Review by SimoneDutto — COMMENTED** [review]
few comments then LGTM

File: internal/jimm/jimm.go (hunk 20 above)
Comment (alesstimec) [reply internal/jimm/jimm.go]: we probably could.. but then we'd need to provide a "valid" set of parameters every time constructor was called

File: internal/jimm/jimm.go [hunk 21]
```diff
@@ -41,16 +42,103 @@ var (
...
+	CountRoles(ctx context.Context, user *openfga.User) (int, error)
+}
+
+// Option modifies the JIMM returned by the constructor.
+// This is useful to create mock domain objects, e.g. replace the real RoleManager with a mock
+// object in tests.
+type Option func(*JIMM)
+
+// RoleManagerOption replaces the role manager in the returned JIMM with the specified one.
+var RoleManagerOption = func(rm RoleManager) Option {
```
Comment (ale8k) [inline internal/jimm/jimm.go]: Just make a func called WithRoleManager?

File: internal/jimm/jimm.go [hunk 22]
```diff
@@ -87,28 +175,86 @@ type JIMM struct {
...
+
+	if p.OpenFGAClient == nil {
+		return errors.E("missing openfga client")
+	}
+
+	if p.JWKService == nil {
+		return errors.E("missing jwks service")
+	}
+
+	if p.JWTService == nil {
```
Comment (ale8k) [inline internal/jimm/jimm.go:217]: I imagine this should only contain primitives? Not Serices?

File: internal/jimm/jimm.go (hunk 8 above)
Comment (ale8k) [reply internal/jimm/jimm.go:265]: I wouldn't embed it, params you pass into the func, and on JIMM it filles in another struct. Don't couple them please?

File: internal/jimm/jimm.go (hunk 20 above)
Comment (alesstimec) [reply internal/jimm/jimm.go]: done

File: internal/jimm/jimm.go (hunk 21 above)
Comment (ale8k) [reply internal/jimm/jimm.go]: Also uhm, might be a good idea to have a file just called options? Be easier to track these options as time goes on.

Comment (alesstimec) [reply internal/jimm/jimm.go]: removed

File: cmd/jimmsrv/service/service.go (hunk 13 above)
Comment (alesstimec) [reply cmd/jimmsrv/service/service.go:559]: improve how?

File: internal/jimm/access_test.go [hunk 23]
```diff
@@ -428,20 +415,7 @@ func TestParseAndValidateTag(t *testing.T) {
...
-		UUID: uuid.NewString(),
-		Database: db.Database{
-			DB: jimmtest.PostgresDB(c, func() time.Time { return now }),
-		},
-		OpenFGAClient: ofgaClient,
-	}
-
-	err = j.Database.Migrate(ctx, false)
-	c.Assert(err, qt.IsNil)
+	j := jimmtest.NewJIMM(c, nil)
```
Comment (SimoneDutto) [inline internal/jimm/access_test.go:418]: this is very nice!

File: internal/testutils/jimmtest/jimm.go [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	if p.CredentialStore == nil {
+		p.CredentialStore = p.Database
+	}
+	if p.OpenFGAClient == nil {
+		ofgaClient, _, _, err := SetupTestOFGAClient(t.Name())
+		if err != nil {
+			t.Fatalf("setting up openfga client: %v", err)
+		}
+		p.OpenFGAClient = ofgaClient
```
Comment (SimoneDutto) [inline internal/testutils/jimmtest/jimm.go:85]: so, this is the default config for jimmtest now. Update the godoc please

File: internal/testutils/jimmtest/jimm.go [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+type Option func(j *jimm.JIMM)
+
+var (
+	UnsetCredentialStore Option = func(j *jimm.JIMM) {
+		j.CredentialStore = nil
+	}
+)
+
+func NewJIMM(t Tester, additionalParameters *jimm.Parameters, options ...Option) *jimm.JIMM {
```
Comment (SimoneDutto) [inline internal/testutils/jimmtest/jimm.go:26]: godoc

File: internal/testutils/jimmtest/jimm.go [hunk 26]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2024 Canonical.
+
```
Comment (SimoneDutto) [inline internal/testutils/jimmtest/jimm.go:2]: i was hoping for my first "new line please". I'll get you next time

Comment (SimoneDutto) [inline internal/testutils/jimmtest/jimm.go:2]: i was hoping for my first "new line please". I'll get you next time

File: internal/testutils/jimmtest/jimm.go [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+var now = (time.Time{}).UTC().Round(time.Millisecond)
+
+type Option func(j *jimm.JIMM)
+
+var (
+	UnsetCredentialStore Option = func(j *jimm.JIMM) {
+		j.CredentialStore = nil
+	}
+)
```
Comment (SimoneDutto) [inline internal/testutils/jimmtest/jimm.go:24]: if we give default options to choose from, do we need to have the `Option` exported?

File: internal/testutils/jimmtest/suite.go [hunk 28]
```diff
@@ -47,7 +46,11 @@ func (t GocheckTester) Name() string {
 }
 
 func (t GocheckTester) Cleanup(f func()) {
-	t.C.Logf("warning: gocheck does not support Cleanup functions; make sure you're using suite's tear-down method")
+	if t.AddCleanup != nil {
+		t.AddCleanup(f)
```
Comment (SimoneDutto) [inline internal/testutils/jimmtest/suite.go:50]: why bother with adding the cleanup? Since we don't have any cleanup with all the `ŧesting.t` tests.

I think it's just enough to increase the conn limit on postgres

Comment (SimoneDutto) [reply internal/testutils/jimmtest/suite.go:50]: btw, where did the change go? Or now it's fine because we are not creating that many connections?

**Review by SimoneDutto — APPROVED** [review]
lgtm, but please new time let's split this in multiple pr, with a feature branch. 
Too many mechanical changes in the same pr

File: internal/testutils/jimmtest/jimm.go (hunk 27 above)
Comment (alesstimec) [reply internal/testutils/jimmtest/jimm.go:24]: yes. we have a slight bug around cloud credentials. and there is a test that relies on the fact that the credential store is unset.. until we fix that bug, we need this option

File: internal/testutils/jimmtest/suite.go (hunk 28 above)
Comment (alesstimec) [reply internal/testutils/jimmtest/suite.go:50]: https://github.com/canonical/jimm/pull/1468/files#diff-3fde9d1a396e140fefc7676e1bd237d67b6864552b6f45af1ebcc27bcd0bb6e9R73

File: cmd/jimmsrv/service/service.go [hunk 29]
```diff
@@ -389,42 +377,36 @@ func NewService(ctx context.Context, p Params) (*Service, error) {
...
+		jimmParameters.Dialer = jimm.CacheDialer(jimmParameters.Dialer)
 	}
 
 	if _, err := url.Parse(p.DashboardFinalRedirectURL); err != nil {
 		return nil, errors.E(op, err, "failed to parse final redirect url for the dashboard")
 	}
 
-	rebacBackend, err := rebac_admin.SetupBackend(ctx, &s.jimm)
+	// instantiate jimm
+	s.jimm, err = jimm.New(jimmParameters)
```
Comment (ale8k) [inline cmd/jimmsrv/service/service.go:400]: Happy with this as is, as we discussed, but we should definitely break this down further into primitives and application domains later on. I'd love to see primitive db, openfga, authsvc (as params) -> and within the constructor: rolemanager, identityservice, etc. all constructed returning a nicely complete jimm. And, if we want to override any application domain we pass an option.

File: cmd/jimmsrv/service/service.go [hunk 30]
```diff
@@ -486,16 +473,69 @@ func NewService(ctx context.Context, p Params) (*Service, error) {
...
-	s.mux.Handle("/model/*", websocketCors.Handler(http.StripPrefix("/model", jujuapi.ModelHandler(ctx, &s.jimm, params))))
+	s.mux.Handle("/api", websocketCors.Handler(jujuapi.APIHandler(ctx, s.jimm, params)))
+	s.mux.Handle("/model/*", websocketCors.Handler(http.StripPrefix("/model", jujuapi.ModelHandler(ctx, s.jimm, params))))
 	mountHandler(
 		"/model/{uuid}/{type:charms|applications}",
-		jimmhttp.NewHTTPProxyHandler(&s.jimm),
+		jimmhttp.NewHTTPProxyHandler(s.jimm),
 	)
 
+	if p.AuditLogRetentionPeriodInDays != "" {
```
Comment (ale8k) [inline cmd/jimmsrv/service/service.go:483]: I feel like we should have service params honestly, it's confusing / bit odd, to see params which re for both jimm and the service and they take effect when you start the service. @kian99 I think you mentioned this right?

File: cmd/jimmsrv/service/service.go [hunk 31]
```diff
@@ -486,16 +473,69 @@ func NewService(ctx context.Context, p Params) (*Service, error) {
...
+			return nil, errors.E(op, "retention period cannot be less than 0")
+		}
+	}
+	s.isLeader = p.IsLeader
+
 	return s, nil
 }
 
+func (s *Service) StartServices(ctx context.Context, svc *service.Service) {
+	// on the leader unit we start additional routines
```
Comment (ale8k) [inline cmd/jimmsrv/service/service.go:499]: I do like how we start all services here, but in main.go we're running this and then we start the jimm service. Would be nice if jimm was also started here? Could even setup the on shutdown etc all here.

File: internal/db/cloud_test.go [hunk 32]
```diff
@@ -244,7 +244,7 @@ controllers:
     region: test-region
     priority: 1
 `)
-	env.PopulateDB(c, *s.Database)
+	env.PopulateDB(c, s.Database)
```
Comment (ale8k) [inline internal/db/cloud_test.go:247]: Interesting we could handle ** here. Good spot.

File: cmd/jimmsrv/service/service.go (hunk 13 above)
Comment (kian99) [reply cmd/jimmsrv/service/service.go:559]: For a separate PR, and not for you per se, but I was just remarking that the naming isn't great. database.database.database() is crazy.

File: docker-compose.yaml [hunk 33]
```diff
@@ -70,7 +70,7 @@ services:
       POSTGRES_PASSWORD: jimm
     # Since it's mainly used for testing purposes, it's okay to set fsync=off for
     # improved performance.
-    command: -c fsync=off -c full_page_writes=off
+    command: -c fsync=off -c full_page_writes=off -c max_connections=200
```
Comment (kian99) [inline docker-compose.yaml:73]: This is a bit concerning. I wonder if we're going over the default 100 connections because we're leaking connections or because of tests running in parallel. Perhaps one way to check would be to add a global test cleanup methods that checks the number of open connections to Postgres. Or run tests with -p 1 and print the number of connections to Postgres between each test incovation.

File: internal/jimm/jimm.go (hunk 8 above)
Comment (kian99) [reply internal/jimm/jimm.go:265]: They're only embedded for the time being. Once we have the domain objects, we can delete this.

File: internal/testutils/jimmtest/jimm.go [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+			p.JWTService = additionalParameters.JWTService
+		}
+		if additionalParameters.OAuthAuthenticator != nil {
+			p.OAuthAuthenticator = additionalParameters.OAuthAuthenticator
+		}
+	}
+
+	if p.Database == nil {
+		p.Database = &db.Database{
+			DB: PostgresDB(t, func() time.Time { return now }),
```
Comment (kian99) [inline internal/testutils/jimmtest/jimm.go:74]: I'm currently not in favour of this approach where every test that calls `jimmtest.NewJIMM` gets a whole database and OpenFGA store. Some tests need neither, some tests need one. That was nicely captured in the tests and when we move to managers we can use suites so that there is less duplication of that boilerplate test setup.

I think the default should be all mocks or all nil fields unless the test requests them.

File: internal/testutils/jimmtest/suite.go [hunk 35]
```diff
@@ -172,14 +190,14 @@ func (s *JIMMSuite) TearDownTest(c *gc.C) {
 	}
 }
 
-func (s *JIMMSuite) setupMacaroonDischarger(c *gc.C) *discharger.MacaroonDischarger {
+func setupMacaroonDischarger(c *gc.C, uuid string, db *db.Database, ofgaClient *openfga.OFGAClient) *discharger.MacaroonDischarger {
```
Comment (kian99) [inline internal/testutils/jimmtest/suite.go:193]: Why was this changed to a function?

**Review by kian99 — COMMENTED** [review]
I think it's almost okay. My biggest gripe is with all tests getting a DB and OpenFGA client.

File: internal/testutils/jimmtest/suite.go (hunk 28 above)
Comment (kian99) [reply internal/testutils/jimmtest/suite.go:50]: I think this link stops working when you force push btw.

File: internal/testutils/jimmtest/jimm.go [hunk 36]
```diff
@@ -0,0 +1,98 @@
...
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/jimm"
+	"github.com/canonical/jimm/v3/internal/jimmjwx"
+	"github.com/canonical/jimm/v3/internal/pubsub"
+)
+
+var now = (time.Time{}).UTC().Round(time.Millisecond)
+
+type Option func(j *jimm.JIMM)
```
Comment (ale8k) [inline internal/testutils/jimmtest/jimm.go:18]: I see, we use option pattern only in the context of tests... I think... It's more useful the have the pattern on the JIMM constructor for when we come to (or not to) initialise services. Kinda like juju's manifolds where they configure the binary in different ways - if we ever need to go down that path we can. It's also no harm I think rather than having the option pattern only apply to the jimmtest constructor, I think it's a lot clearer if we just use jimm.New() and pass test options.

**Review by ale8k — APPROVED** [review]

**Review by ale8k — APPROVED** [review]
lgtm, I've left some comments. Nothing too serious, but more about options and service. I'd much prefer we just use a single constructor for initialisng jimm in tests and in main.

File: internal/testutils/jimmtest/jimm.go (hunk 36 above)
Comment (alesstimec) [reply internal/testutils/jimmtest/jimm.go:18]: nope.. in  production we never expect to use options.. only for tests, so why pollute production code with options?
